query-engine/prisma/src/tests/test_api.rs

```diff
@@ -0,0 +1,269 @@
+    PrismaRequest, PrismaResponse,
+};
+use migration_connector::*;
+use migration_core::{
+    api::{GenericApi, MigrationApi},
+    commands::{ApplyMigrationInput, InferMigrationStepsInput, ResetCommand},
+};
+use quaint::{
+    ast::*,
+    connector::ConnectionInfo,
+    visitor::{self, Visitor},
+};
+use sql_migration_connector::SqlMigrationConnector;
+use std::{collections::HashMap, sync::Arc};
+use test_setup::*;
+
+pub struct QueryEngine {
+    context: Arc<PrismaContext>,
+}
+
+impl QueryEngine {
+    pub async fn request(&self, body: impl Into<SingleQuery>) -> serde_json::Value {
+        let request = PrismaRequest {
+            body: GraphQlBody::Single(body.into()),
+            headers: HashMap::new(),
+            path: String::new(),
+        };
+
+        match GraphQlRequestHandler.handle(request, &self.context).await {
+            PrismaResponse::Single(response) => serde_json::to_value(response).unwrap(),
+            _ => unreachable!(),
+        }
+    }
+}
+
+pub struct TestApi {
+    connection_info: ConnectionInfo,
+    migration_api: MigrationApi<sql_migration_connector::SqlMigrationConnector, sql_migration_connector::SqlMigration>,
+    config: String,
+    is_pgbouncer: bool,
+}
+
+impl TestApi {
+    pub async fn create_engine(&self, datamodel: &str) -> anyhow::Result<QueryEngine> {
+        let datamodel_string = format!("{}\n\n{}", self.config, datamodel);
+        let migration_id = "test-cli-migration".to_owned();
+
+        let infer_input = InferMigrationStepsInput {
+            assume_applied_migrations: Some(Vec::new()),
+            assume_to_be_applied: Some(Vec::new()),
+            datamodel: datamodel_string.clone(),
+            migration_id: migration_id.clone(),
+        };
+
+        self.migration_api.reset(&serde_json::Value::Null).await?;
+        let result = self.migration_api.infer_migration_steps(&infer_input).await?;
+
+        let apply_input = ApplyMigrationInput {
+            force: Some(true),
+            migration_id,
+            steps: result.datamodel_steps,
+        };
+
+        self.migration_api.apply_migration(&apply_input).await?;
+
+        let context = PrismaContext::builder()
+            .enable_raw_queries(true)
+            .datamodel(datamodel_string)
+            .force_transactions(self.is_pgbouncer)
+            .build()
+            .await
+            .unwrap();
+
+        Ok(QueryEngine {
+            context: Arc::new(context),
+        })
+    }
+
+    pub fn connection_info(&self) -> &ConnectionInfo {
+        &self.connection_info
+    }
+
+    pub fn to_sql_string<'a>(&'a self, query: impl Into<Query<'a>>) -> (String, Vec<ParameterizedValue>) {
+        match self.connection_info() {
+            ConnectionInfo::Postgres(..) => visitor::Postgres::build(query),
+            ConnectionInfo::Mysql(..) => visitor::Mysql::build(query),
+            ConnectionInfo::Sqlite { .. } => visitor::Sqlite::build(query),
+        }
+    }
```

Comment:
that's super nice!

Reply:
Only for execute raw needs though. But still yeah I like it.